onnxruntime/core/graph/contrib_ops/nhwc_inference_context.h

```diff
@@ -16,113 +16,94 @@ namespace contrib {
-
-    // copy latest type/shape info.
-    nhwc_tp = output_type_;
-
-    // convert shape to channels last
-    if (output_type_.tensor_type().has_shape()) {
-      const auto& nchw_shape = output_type_.tensor_type().shape();
-      const int rank = nchw_shape.dim_size();
-      // N and C dims are required. Some operators like AveragePool allow 1D input
-      if (rank < 3) {
-        fail_shape_inference("Output tensor must have at least 3 dimensions");
-      }
-
-      // Convert output shape from N, C, H {, W, ...} to N, H {, W, ...}, C.
-      auto& nhwc_shape = *nhwc_tp.mutable_tensor_type()->mutable_shape();
-      nhwc_shape.Clear();
-      *nhwc_shape.add_dim() = nchw_shape.dim(0);
-      for (int i = 2; i < rank; i++) {
-        *nhwc_shape.add_dim() = nchw_shape.dim(i);
-      }
-
-      *nhwc_shape.add_dim() = nchw_shape.dim(1);
-    }
-  }
+  void PropagateOutputShape();
 
  private:
-  void TransposeToNchw(const ONNX_NAMESPACE::TypeProto& nhwc_tp, ONNX_NAMESPACE::TypeProto& nchw_tp) {
-    if (nhwc_tp.tensor_type().has_shape()) {
-      const auto& nhwc_shape = nhwc_tp.tensor_type().shape();
-      const int rank = nhwc_shape.dim_size();
-      // N and C dims are required. Some operators like AveragePool allow 1D input.
-      if (rank < 3) {
-        fail_shape_inference(
-            "Tensor must have at least 3 dimensions to convert between channels first and channels last.");
-      }
-
-      // Convert input shape from {N, H, W, ..., C} to {N, C, H, W, ...}.
-      auto& nchw_shape = *nchw_tp.mutable_tensor_type()->mutable_shape();
-      nchw_shape.Clear();
-      *nchw_shape.add_dim() = nhwc_shape.dim(0);
-      *nchw_shape.add_dim() = nhwc_shape.dim(rank - 1);
-      for (int i = 1; i < rank - 1; i++) {
-        *nchw_shape.add_dim() = nhwc_shape.dim(i);
-      }
-    }
-  }
+  InferenceContext& ctx_;
+
+  // These are the NCHW versions of the actual input and output types.
+  // Provided to ONNX type/shape inferencing via the overridden InferenceContext::getInputType function.
+  ONNX_NAMESPACE::TypeProto input_type_;
+  ONNX_NAMESPACE::TypeProto output_type_;
+};
+
+/** Adapter class to enable ONNX shape inferencing to be used with the NHWC Resize layout-sensitive operator.
+ * Currently only used by the QNN EP.
+ *
+ * After layout transformation, the Resize operator's input0 and output0 have an NHWC layout. However, ONNX shape
+ * inferencing expects a NCHW layout for inputs and outputs. This adapter ensures that ONNX sees a NCHW version
+ * of the operator during shape inferencing.
+ *
+ * The NhwcInferenceContext class is not used for NHWC Resize because it only handles input0 and output0. We also
+ * need to convert the 'scales' and 'sizes' constant inputs to their NCHW counterparts.
+ * Ex: a NHWC 'scales' input [1.0f, 2.0f, 3.0f, 4.0f] needs to be converted to [1.0f, 4.0f, 2.0f, 3.0f] for NCHW.
+ */
+class NhwcResizeInferenceContext : public ONNX_NAMESPACE::InferenceContext {
```

Comment:
Good point. Assuming this the general direction in which we want to proceed, I'll modify this class to inherit from `NhwcInferenceContext` so that we can more easily leverage common functionality.